lib/psci/psci_suspend.c

```diff
@@ -85,6 +85,14 @@ static void psci_suspend_to_pwrdown_start(unsigned int end_pwrlvl,
 	 */
 	cm_init_my_context(ep);
 
+	/*
+	 * Plat. management: Allow the platform to perform any early
+	 * actions required to power down the CPU, before we disable
+	 * the data cache.
+	 */
+	if (psci_plat_pm_ops->pwr_domain_suspend_early)
+		psci_plat_pm_ops->pwr_domain_suspend_early(state_info);
+
```

Comment:
I had a chat with Soby about this and we agreed the easiest thing is to go his way. So that means:

*   Move this call to just after `psci_spd_pm->svc_suspend()`. At the same time this can be wrapped in `#if !HW_ASSISTED_COHERENCY` to make it clear this is only useful in this case.
*   Rename the function to `pwr_domain_suspend_pwrdown_early()`.
*   Document the semantics of the new optional hook properly. Ignore my comments about merging `pwr_domain_suspend()` and `pwr_domain_suspend_early()` documentation. However, it should be possible to for the new function to refer to the existing `pwr_domain_suspend()` documentation rather than copy/pasting.